openstate/open-cultuur-data
# Usage logging

This pull-request adds usage logging functionality to the Open Cultuur Data API. The following events are captured with the code that is included within this pull-request:
* ``search``
* ``search_similar``
* ``sources``
* ``get_object``
* ``get_object_source``
* ``resolve``

Events are stored as separate documents in an Elasticsearch index. To minimize the impact of logging on the response time of the REST API, actual storage and formatting is taken care of asynchronously by Celery tasks. The raw logs aren't made publicly available by the API. Instead I implemented a very simple ``stats`` endpoint that allows users to get some data about a single object. We can expand this functionality once we have a better idea of the kind of insights are provided by the logs (i.e. after we've collected some data). Meanwhile Kibana or 'raw' ES quereis could be used to allow OCD staff to explore the logs.

Testing the logging functionality
------------------------------------------
1. Create the OCD with the associated schema's via the CLI command (``./manage.py elasticsearch create_indexes es_mappings``)
2. Run a Celery worker to process incoming log requests (``celery --app="ocd_frontend.rest.tasks:celery" worker --loglevel=info --concurrency=2``)
3. Start the REST API (``./manage.py frontend runserver``)
4. Hit some endpoints and check if the events appear in the Elasticsearch index

By default we use Redis db 1, but obviously this can be changed in the ``ocd_frontend`` settings.

Other changes and fixes
---------------------------------
* I refactored the frontend tests to better split-up tests of different endpoints
* All endpoints now contain basic tests
* While refactoring the tests I came across some bugs, fixed those
* Some small fixes to different parts of the documentation
* Made some changes to the bootstrapping process of vagrant boxes (install ES head plugin, fixed ``bashrc``, etc.)
